### PR TITLE
🛠 Fix timespan from date expr

### PR DESCRIPTION
**Pull request by AyhamAl-Ali**
### Description

This PR fixes a bug explained in this issue #4431

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:**  #4431

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 1]
```diff
@@ -45,28 +45,37 @@
...
-    @SuppressWarnings("null")
-    @Nullable
     @Override
+    @Nullable
+    @SuppressWarnings("null")
     protected Date[] get(Event e) {
         Timespan timespan = this.timespan.getSingle(e);
-        Date date = this.date == null ? new Date() : this.date.getSingle(e);
-        if (timespan == null || date == null) {
+        Date date = this.date == null ? new Date() : new Date(this.date.getSingle(e).getTimestamp());
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:69]: You should check if getSingle returns null before calling any methods

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java (hunk 1 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:69]: Thank you for letting me know, I will do that in a moment \:D

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 2]
```diff
@@ -45,28 +45,37 @@
...
 
     private boolean ago;
 
-    @SuppressWarnings("null")
-    @Nullable
     @Override
+    @Nullable
+    @SuppressWarnings("null")
     protected Date[] get(Event e) {
+		if (this.date == null || this.date.getSingle(e) == null || this.timespan == null || this.timespan.getSingle(e) == null)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:68]: I would avoid calling getSingle twice. Perhaps just store the value from getSingle and check whether that is null

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java (hunk 2 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:68]: Sure, is this needed btw `(this.date == null || this.timespan == null)`?

Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java]: date yes, timespan no, see the pattern: the date expression is optional, the timespan isn't

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 3]
```diff
@@ -45,28 +45,41 @@
...
 
     private boolean ago;
 
-    @SuppressWarnings("null")
-    @Nullable
     @Override
+    @Nullable
+    @SuppressWarnings("null")
     protected Date[] get(Event e) {
+		if (this.date == null || this.timespan == null)
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:68]: This'll make the syntax with the optional date part always return null

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 4]
```diff
@@ -45,28 +45,41 @@
...
-        if (timespan == null || date == null) {
-            return null;
-        }
-        if (ago) {
+		Date otherDate = this.date.getSingle(e);
+		if (timespan == null || otherDate == null)
+			return null;
+
+        Date date = this.date == null ? new Date() : new Date(otherDate.getTimestamp());
+        if (ago)
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:77]: You should use Date#plus and Date#minus, it'll remove the need to create a new date instance yourself too

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 5]
```diff
@@ -45,28 +45,41 @@
                 "%timespan% (later|(from|after) [the] [date] %-date%)");
     }
 
+    @Override
+    @SuppressWarnings({"unchecked", "null"})
+    public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:50]: Import ParseResult directly

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 6]
```diff
@@ -83,13 +96,4 @@ public boolean isSingle() {
     public String toString(@Nullable Event e, boolean debug) {
         return timespan.toString(e, debug) + " " + (ago ? "ago" : "later");
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:97]: Can you also fix this? It currently doesn't include the date expr part

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

**Comment by AyhamAl-Ali** [comment]
@TPGamesNL could you also check if I used the new tags correctly, I tested and they seems to work as expected

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java [hunk 7]
```diff
@@ -49,24 +51,30 @@
...
+        ago = matchedPattern == 0;
+        return true;
+    }
+
+    @Override
+    @Nullable
+    @SuppressWarnings("null")
     protected Date[] get(Event e) {
+		if (timespan == null)
+			return null;
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/expressions/ExprDateAgoLater.java:70]: Can be removed, always false

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

**Review by TPGamesNL — APPROVED** [review]

**Pull request merged**